plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClientModule.java

```diff
@@ -42,13 +44,33 @@ public void configure(Binder binder)
         binder.bind(JdbcClient.class).annotatedWith(ForBaseJdbc.class).to(ClickHouseClient.class).in(Scopes.SINGLETON);
         bindTablePropertiesProvider(binder, ClickHouseTableProperties.class);
         binder.install(new DecimalModule());
+        binder.bind(ConnectionFactory.class).annotatedWith(ForBaseJdbc.class).toProvider(ConnectionFactoryProvider.class).in(Scopes.SINGLETON);
     }
 
-    @Provides
-    @Singleton
-    @ForBaseJdbc
-    public static ConnectionFactory createConnectionFactory(BaseJdbcConfig config, CredentialProvider credentialProvider)
+    private static class ConnectionFactoryProvider
+            implements Provider<ConnectionFactory>
     {
-        return new ClickHouseConnectionFactory(new DriverConnectionFactory(new ClickHouseDriver(), config, credentialProvider));
+        private final ConnectionFactory connectionFactory;
+
+        @Inject
+        public ConnectionFactoryProvider(ClickHouseConfig clickHouseConfig, BaseJdbcConfig baseJdbcConfig, CredentialProvider credentialProvider)
+        {
+            connectionFactory = new ClickHouseConnectionFactory(
+                    new DriverConnectionFactory(createDriver(clickHouseConfig.isUseDeprecatedDriver()), baseJdbcConfig, credentialProvider));
+        }
+
+        @Override
+        public ConnectionFactory get()
+        {
+            return connectionFactory;
+        }
+
+        Driver createDriver(boolean useDeprecatedDriver)
+        {
+            if (useDeprecatedDriver) {
+                return new ru.yandex.clickhouse.ClickHouseDriver();
+            }
+            return new com.clickhouse.jdbc.ClickHouseDriver();
+        }
     }
```

Comment:
You don't need to rewrite the method entirely. Please replace `new ClickHouseDriver()` with `createDriver(clickHouseConfig...)`.